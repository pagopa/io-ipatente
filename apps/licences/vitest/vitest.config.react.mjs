/// <reference types="vitest" />
/// <reference types="vite/client" />

import react from "@vitejs/plugin-react";
import { loadEnv } from "vite";
import tsconfigPaths from "vite-tsconfig-paths";
import { configDefaults, defineConfig } from "vitest/config";

export default defineConfig({
  plugins: [react(), tsconfigPaths()],
  test: {
    coverage: {
      exclude: [
        ...configDefaults.coverage.exclude,
        "**/__tests__",
        "src/generated/**",

        // BE
        "src/app/*",
        "src/config/*",
        "src/lib/bff/*",
      ],
      ignoreEmptyLines: true,
      include: ["src/**/*.{tsx,ts}"],
      reporter: ["text", "html"],
    },
    css: true,
    env: loadEnv("test", process.cwd(), ""),
    environment: "jsdom",
    exclude: [
      ...configDefaults.exclude,
      "./vitest/vitest.setup.react.ts",
      "./vitest/vitest.setup.node.ts",
    ],
    globals: true,
    include: ["./src/**/*.test.tsx"],
    setupFiles: "./vitest/vitest.setup.react.ts",
    typecheck: {
      ignoreSourceErrors: true,
    },
  },
});
