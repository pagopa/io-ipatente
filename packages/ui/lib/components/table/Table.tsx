import { Typography } from "@mui/material";
import Paper from "@mui/material/Paper";
import MuiTable from "@mui/material/Table";
import TableBody from "@mui/material/TableBody";
import TableCell from "@mui/material/TableCell";
import TableContainer from "@mui/material/TableContainer";
import TableHead from "@mui/material/TableHead";
import TableRow from "@mui/material/TableRow";
import { CSSProperties, ReactElement } from "react";

type WidthFactor = 0.1 | 0.2 | 0.3 | 0.4 | 0.5 | 0.6 | 0.7 | 0.8 | 0.9;
export interface Column<T> {
  columnStyle?: CSSProperties;
  key: keyof T;
  render?: (key: Column<T>["key"], item: T) => ReactElement;
  title: string;
  widthFactor?: WidthFactor;
}

export interface TableProps<T> {
  columns: Column<T>[];
  rows: T[];
}

export const Table = <T,>({ columns, rows }: TableProps<T>) => (
  <TableContainer component={Paper}>
    <MuiTable aria-label="table">
      <TableHeader columns={columns} />
      <TableContent columns={columns} rows={rows} />
    </MuiTable>
  </TableContainer>
);

interface TableHeaderProps<T> {
  columns: Column<T>[];
}

const TableHeader = <T,>({ columns }: TableHeaderProps<T>) => {
  const columnsWidths = calculateWidth(columns);

  return (
    <TableHead>
      <TableRow>
        {columns.map(({ columnStyle = {}, key, title }, index) => (
          <TableCell
            key={`${key.toString()}-${index}`}
            sx={{
              backgroundColor: "text.primary",
              color: "background.default",
              paddingX: 1.5,
              paddingY: 2,
              width: `${columnsWidths[index] * 100}%`,
              ...columnStyle,
            }}
          >
            {title}
          </TableCell>
        ))}
      </TableRow>
    </TableHead>
  );
};

const TableContent = <T,>({ columns, rows }: TableProps<T>) => (
  <TableBody>
    {rows.map((row, index) => (
      <TableRow
        key={index}
        sx={{
          "&:last-child td, &:last-child th": { border: 0 },
          "&>td, &>th": { borderBottom: "1px solid #CBD5E1" },
        }}
      >
        {columns.map(({ key, render }, cellIndex) => (
          <TableCell
            component="th"
            key={`${key.toString()}-${index}-${cellIndex}`}
            scope="row"
            sx={{
              padding: 1.5,
            }}
          >
            {render ? (
              render(key, row)
            ) : (
              <Typography
                color="text.secondary"
                fontWeight="regular"
                textAlign="initial"
                variant="body1"
              >
                {row[key] as string}
              </Typography>
            )}
          </TableCell>
        ))}
      </TableRow>
    ))}
  </TableBody>
);

function calculateWidth<T>(columns: Column<T>[]) {
  const widths = columns.map(({ widthFactor }) => widthFactor);
  const { emptyCount, sum } = widths.reduce<{
    emptyCount: number;
    sum: number;
  }>(
    ({ emptyCount, sum }, width) =>
      width
        ? { emptyCount, sum: sum + width }
        : { emptyCount: emptyCount + 1, sum },
    { emptyCount: 0, sum: 0 },
  );

  const defaultWidth = emptyCount
    ? Number(((1 - sum) / emptyCount).toFixed(1))
    : Number((1 / columns.length).toFixed(1));
  return widths.map((width) => (width ? width : defaultWidth));
}
