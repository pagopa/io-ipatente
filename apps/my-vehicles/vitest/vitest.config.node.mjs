import { configDefaults, defineConfig } from "vitest/config";
import tsconfigPaths from "vite-tsconfig-paths";

export default defineConfig({
  plugins: [tsconfigPaths()],
  test: {
    coverage: {
      exclude: [
        ...(configDefaults.coverage.exclude
          ? configDefaults.coverage.exclude
          : []),
        ".next/*",
        "mocks/*",
        "public/*",
        "*.config.js",
        "src/generated/**",
        "src/config/*",
        "src/types/*",
        "src/main.ts",
        "src/config.ts",
        "src/instrumentation.ts",
        "**/__mocks__/**",
        // FE
        "src/hooks/*",
        "src/layouts/*",
      ],
      extension: configDefaults.coverage.extension
        ?.toString()
        .replace(",.tsx", "")
        .split(","),
      ignoreEmptyLines: true,
      reporter: ["text", "json-summary", "json"],
      reportsDirectory: `${configDefaults.coverage.reportsDirectory}/backend`,
    },
    exclude: [
      ...configDefaults.exclude,
      "./vitest/vitest.setup.react.ts",
      "./vitest/vitest.setup.node.ts",
    ],
    include: ["./src/**/__tests__/*.ts"],
    setupFiles: ["./vitest/vitest.setup.node.ts"],
    typecheck: {
      ignoreSourceErrors: true,
    },
  },
});
