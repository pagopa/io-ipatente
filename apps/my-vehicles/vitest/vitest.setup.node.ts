/**
 * MSW configuration for Vitest (used as setupFile in vitest.config.node.ts)
 */
import { loadEnvConfig } from "@next/env";
import { afterAll, afterEach, beforeAll } from "vitest";

import { mswServerTest } from "../mocks/msw-server";

let mswServer: ReturnType<typeof mswServerTest>;
// Start server before all tests
beforeAll(() => {
  loadEnvConfig(process.cwd());
  // this done to correctly load the environment variables on msw handlers
  mswServer = mswServerTest();
  mswServer.listen({ onUnhandledRequest: "warn" });
});

//  Close server after all tests
afterAll(() => mswServer.close());

// Reset handlers after each test `important for test isolation`
afterEach(() => mswServer.resetHandlers());
