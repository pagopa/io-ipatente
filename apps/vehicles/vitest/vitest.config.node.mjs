import { loadEnv } from "vite";
import tsconfigPaths from "vite-tsconfig-paths";
import { configDefaults, defineConfig } from "vitest/config";

export default defineConfig({
  plugins: [tsconfigPaths()],
  test: {
    coverage: {
      exclude: [
        ...configDefaults.coverage.exclude,
        "**/__tests__",
        "src/generated/**",
        "src/app/api/auth/mocks/**",
        // FE
        "src/common/*",
        "src/hooks/*",
      ],
      ignoreEmptyLines: true,
      include: ["src/**/*.ts"],
      reporter: ["text", "html"],
      reportsDirectory: `${configDefaults.coverage.reportsDirectory}/backend`,
    },
    env: loadEnv("test", process.cwd(), ""),
    exclude: [
      ...configDefaults.exclude,
      "./vitest/vitest.setup.react.ts",
      "./vitest/vitest.setup.node.ts",
    ],
    include: ["./src/**/__tests__/*.ts"],
    setupFiles: ["./vitest/vitest.setup.node.ts"],
    typecheck: {
      ignoreSourceErrors: true,
    },
  },
});
