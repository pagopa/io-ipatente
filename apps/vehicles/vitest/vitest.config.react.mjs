/// <reference types="vitest" />
/// <reference types="vite/client" />

import react from "@vitejs/plugin-react";
import tsconfigPaths from "vite-tsconfig-paths";
import { configDefaults, defineConfig } from "vitest/config";
import { loadEnv } from "vite";

export default defineConfig({
  plugins: [react(), tsconfigPaths()],
  test: {
    coverage: {
      exclude: [
        ...(configDefaults.coverage.exclude
          ? configDefaults.coverage.exclude
          : []),
        ".next/*",
        "mocks/*",
        "public/*",
        "*.config.js",
        "src/generated/**",
        "src/config/*",
        "src/types/*",
        "src/main.ts",
        "src/config.ts",
        "src/instrumentation.ts",
        "**/__mocks__/**",
        // BE
        "src/app/*",
        "src/lib/be/*",
      ],
      ignoreEmptyLines: true,
      reporter: ["text", "json-summary", "json"],
      reportsDirectory: `${configDefaults.coverage.reportsDirectory}/frontend`,
    },
    css: true,
    env: loadEnv("test", process.cwd(), ""),
    environment: "jsdom",
    exclude: [
      ...configDefaults.exclude,
      "./vitest/vitest.setup.react.ts",
      "./vitest/vitest.setup.node.ts",
    ],
    globals: true,
    include: ["./src/**/*.test.tsx"],
    setupFiles: "./vitest/vitest.setup.react.ts",
    typecheck: {
      ignoreSourceErrors: true,
    },
  },
});
