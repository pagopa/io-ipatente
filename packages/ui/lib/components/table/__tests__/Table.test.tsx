import { render } from "@testing-library/react";
import { describe, expect, it } from "vitest";

import { Table, TableProps } from "../Table";

const defaultProps: TableProps = {
  columns: ["header-1", "header-2"],
  rows: [
    {
      date: "01/01/1970",
      detail: "Test",
      key: "Test-01/01/1970",
      variation: 3,
    },
    {
      date: "02/01/1970",
      detail: "Test_Error",
      key: "Test_Error-02/01/1970",
      variation: -3,
    },
  ],
};

describe("Test SectionTitle Components", () => {
  it("Should match the snapshot with default props", () => {
    const comp = render(<Table {...defaultProps} />);
    expect(comp).toMatchSnapshot();
  });
});
