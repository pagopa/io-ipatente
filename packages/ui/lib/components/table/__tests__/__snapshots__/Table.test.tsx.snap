// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Test Table Components > Should match the snapshot with default props 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiTableContainer-root css-11xur9t-MuiPaper-root-MuiTableContainer-root"
      >
        <table
          aria-label="table"
          class="MuiTable-root css-rqglhn-MuiTable-root"
        >
          <thead
            class="MuiTableHead-root css-15wwp11-MuiTableHead-root"
          >
            <tr
              class="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
            >
              <th
                class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-pw9qyu-MuiTableCell-root"
                scope="col"
              >
                test_1
              </th>
              <th
                class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-af9c8g-MuiTableCell-root"
                scope="col"
              >
                test_2
              </th>
            </tr>
          </thead>
          <tbody
            class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
          >
            <tr
              class="MuiTableRow-root css-kjco9e-MuiTableRow-root"
            >
              <th
                class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-3jgu42-MuiTableCell-root"
                scope="row"
              >
                <p
                  class="MuiTypography-root MuiTypography-body1 css-crn0fq-MuiTypography-root"
                >
                  Test String
                </p>
              </th>
              <th
                class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-3jgu42-MuiTableCell-root"
                scope="row"
              >
                <p
                  class="MuiTypography-root MuiTypography-body1 css-crn0fq-MuiTypography-root"
                >
                  10
                </p>
              </th>
            </tr>
            <tr
              class="MuiTableRow-root css-kjco9e-MuiTableRow-root"
            >
              <th
                class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-3jgu42-MuiTableCell-root"
                scope="row"
              >
                <p
                  class="MuiTypography-root MuiTypography-body1 css-crn0fq-MuiTypography-root"
                >
                  Test String 2
                </p>
              </th>
              <th
                class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-3jgu42-MuiTableCell-root"
                scope="row"
              >
                <p
                  class="MuiTypography-root MuiTypography-body1 css-crn0fq-MuiTypography-root"
                >
                  20
                </p>
              </th>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiTableContainer-root css-11xur9t-MuiPaper-root-MuiTableContainer-root"
    >
      <table
        aria-label="table"
        class="MuiTable-root css-rqglhn-MuiTable-root"
      >
        <thead
          class="MuiTableHead-root css-15wwp11-MuiTableHead-root"
        >
          <tr
            class="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
          >
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-pw9qyu-MuiTableCell-root"
              scope="col"
            >
              test_1
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-af9c8g-MuiTableCell-root"
              scope="col"
            >
              test_2
            </th>
          </tr>
        </thead>
        <tbody
          class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
        >
          <tr
            class="MuiTableRow-root css-kjco9e-MuiTableRow-root"
          >
            <th
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-3jgu42-MuiTableCell-root"
              scope="row"
            >
              <p
                class="MuiTypography-root MuiTypography-body1 css-crn0fq-MuiTypography-root"
              >
                Test String
              </p>
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-3jgu42-MuiTableCell-root"
              scope="row"
            >
              <p
                class="MuiTypography-root MuiTypography-body1 css-crn0fq-MuiTypography-root"
              >
                10
              </p>
            </th>
          </tr>
          <tr
            class="MuiTableRow-root css-kjco9e-MuiTableRow-root"
          >
            <th
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-3jgu42-MuiTableCell-root"
              scope="row"
            >
              <p
                class="MuiTypography-root MuiTypography-body1 css-crn0fq-MuiTypography-root"
              >
                Test String 2
              </p>
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-3jgu42-MuiTableCell-root"
              scope="row"
            >
              <p
                class="MuiTypography-root MuiTypography-body1 css-crn0fq-MuiTypography-root"
              >
                20
              </p>
            </th>
          </tr>
        </tbody>
      </table>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
